### DISCUSSION TITLE
How to build an arguments method validator properly?

### DISCUSSION THREAD

**Discussion by jeusdi** [making/yavi#380] [Q&A]
I'm trying to build arguments method validator. The method is:

```java
@ConstraintArguments
public PushDocumentResult push(String contentMD5) {
	ArgumentsValidatorBuilder.of(this::push)
		.builder(b -> b.constraint(_PushDocumentUseCasePushArgumentsMeta.CONTENTMD5, c -> c.notEmpty())
	);
}
```

I'm getting this compilation error message:

> The method constraint(StringConstraintMeta<Arguments1<String>>, Function<CharSequenceConstraint<Arguments1<String>,String>,CharSequenceConstraint<Arguments1<String>,String>>) in the type ValidatorBuilder<Arguments1<String>> is not applicable for the arguments (StringConstraintMeta<Arguments2<PushDocumentUseCase,String>>, (<no type> c) -> {})

I don't quite figure out what I'm doing wrong...

Any ideas?

**Comment by making** [comment]
Please provide a link to the runnable example project instead of a snippet of code.

**Reply by jeusdi** [reply]
I can provide full class method. I think it's enought to debug above behavior:

```java
import am.ik.yavi.arguments.Arguments1;
import am.ik.yavi.builder.ArgumentsValidatorBuilder;
import am.ik.yavi.builder.ArgumentsValidatorBuilder.Arguments1ValidatorBuilder;
import am.ik.yavi.builder.ValidatorBuilder;
import am.ik.yavi.meta.ConstraintArguments;

public class PushDocumentUseCase {

	@ConstraintArguments
	public PushDocumentResult push(String contentMD5) {
		ArgumentsValidatorBuilder.of(this::push)
		.builder(b -> b.constraint(_PushDocumentUseCasePushArgumentsMeta.CONTENTMD5,
		c -> c.notEmpty()));
	}
}
```

**Reply by making** [reply]
It's not enough for me to give it a try. `PushDocumentResult` is missing.

**Reply by making** [reply]
Read the document 
* https://yavi.ik.am/#validating-method-arguments-2
* https://yavi.ik.am/#validating-method-arguments

In principle, it is impossible to use automatically generated code in a method with `@ConstraintArguments` that is the target of automatic generation.

**Reply by jeusdi** [reply]
I thought arguments validator stands for validate input parameters of related method.

I see, documentation suggests to use a static validator as a facade of delegated method.
This approach requires a bit more boilerplate of exposed methods, a validator method and real method.

Is this approach valid...?

I mean, I did guess any validated code like:

```
void createCustomer(String name, String address, Integer age, String secondName) {
  //validate here input parameters...
}
```

**Reply by making** [reply]
You can use the argumentes validator inside the method. (calling `lazy()` may be needed to avoid stackoverflow)
But in this case, you can't use `@ConstraintArguments` as explained above.

Anyway if you want me to give it a try, again, please provide the runnable (minimal) example project

**Reply by jeusdi** [reply]
Perhaps I'm not fully grasping a key aspect that's preventing me from unlocking my coding proposal.

From what you've mentioned, I gather that I might not be framing the code correctly, but I'm struggling to see how to approach it differently.

I've provided an example of standalone code that isn't reliant on any structure. How would you suggest approaching the code?

```java
@ConstraintArguments
public String pushOne(String contentMD5) {
	Arguments1Validator<String, String> arguments1Validator = ArgumentsValidatorBuilder.of(this::pushOne)
			.builder(b -> b.constraint(_PushDocumentUseCasePushOneArgumentsMeta.CONTENTMD5, c -> c.notBlank()))
			.build();

	Validated<String> validate = arguments1Validator.validate(contentMD5);
	if (validate.isValid()) {
         return "ok";
	}
	return "ko";
}
```

**Reply by making** [reply]
> Anyway if you want me to give it a try, again, please provide the runnable (minimal) example project

**Reply by jeusdi** [reply]
Here you have it: https://github.com/jcabreroca/yavi-issue

**Reply by making** [reply]
The below should work. 
As I explained, you can't use the validator that uses the generated `ConstraintMeta` in the class with `@ConstraintArguments`.
So you need to call the validator from a different class.

```java
import am.ik.yavi.arguments.Arguments2Validator;
import am.ik.yavi.builder.ArgumentsValidatorBuilder;
import am.ik.yavi.meta.ConstraintArguments;

public class Foo {
	public static Arguments2Validator<Foo, String, String> pushOneValidator = ArgumentsValidatorBuilder.of(Foo::pushOne)
			.builder(b -> b.constraint(_FooPushOneArgumentsMeta.CONTENTMD5, c -> c.notBlank()))
			.build();
	@ConstraintArguments
	public String pushOne(String contentMD5) {
		return contentMD5.toUpperCase();
	}
}

```

```java
import am.ik.yavi.core.ConstraintViolationsException;
import am.ik.yavi.core.Validated;

public class App {
	public static void main(String[] args) {
		Foo foo = new Foo();
		{
			Validated<String> validated = Foo.pushOneValidator.validate(foo, "foo");
			String result = validated.orElseThrow(ConstraintViolationsException::new);
			System.out.println(result); // FOO
		}
		{
			Validated<String> validated = Foo.pushOneValidator.validate(foo, "");
			String result = validated.orElseThrow(ConstraintViolationsException::new); // am.ik.yavi.core.ConstraintViolationsException:  Constraint violations found! (* "contentMD5" must not be blank)
			System.out.println(result);
		}
		// or
		System.out.println(Foo.pushOneValidator.validated(foo, "foo")); // FOO
		System.out.println(Foo.pushOneValidator.validated(foo, "")); // am.ik.yavi.core.ConstraintViolationsException:  Constraint violations found! (* "contentMD5" must not be blank)

	}
}
```

If you want to use the validator inside the method. You can define the validator without `@ConstraintArguments` as follows:

```java
import am.ik.yavi.arguments.Arguments2Validator;
import am.ik.yavi.meta.ConstraintArguments;
import am.ik.yavi.validator.Yavi;

public class Foo {
	public static Arguments2Validator<Foo, String, String> pushOneValidator = Yavi.arguments()
			.<Foo>_object("foo", c -> c.notNull())
			._string("contentMD5", c -> c.notBlank())
			.apply(Foo::pushOne);

	@ConstraintArguments
	public String pushOne(String contentMD5) {
		pushOneValidator.lazy().validated(this, contentMD5);
		return contentMD5.toUpperCase();
	}
}
```

```java
public class App {
	public static void main(String[] args) {
		Foo foo = new Foo();
		System.out.println(foo.pushOne("foo")); // FOO
		System.out.println(foo.pushOne("")); // am.ik.yavi.core.ConstraintViolationsException:  Constraint violations found! (* "contentMD5" must not be blank)
	}
}
```

Note that this easier way to define an arguments validator has been supported since YAVI 0.14.0. https://github.com/making/yavi/pull/367
